toc.md

```diff
@@ -1,54 +1,76 @@
 
 #### Jobs Persistence
+
+As mentioned above, a question that needs to be answered is: how can we make sure our jobs don't get lost if the BJP or the main application crashes?
+To solve this problem we introduced a data store. This data store is used to persist the serialized jobs that have been enqueued by the main application.
+For Workerholic, we decided to use Redis because of its following features:
+- convenient data structures for the problems we needed to solve (lists, sorted sets, hashes)
+- persistence to disk (every 5 minutes by default, configurable)
+- key:value data store, which grants easy access to the data
+- in memory data store, which allows for very efficient reads and writes
+
 ![Jobs Persistence Diagram](/images/jobs_persistence_redis.png)
 
-In order to make our jobs persistent even through some of those failures, we rely on Redis's robustness to make our library more reliable. Redis helps solve the problem of when either the web application crashes or if the background job processor itself crashes, the jobs that are already stored on Redis will be preserved. But what if Redis itself crashes? This is part of why Redis is considered robust because it takes snapshots of your database every five minutes by default. As an added bonus, you can configure Redis to take snapshots more or less as needed.
+By relying on Redis and its robustness we made Workerholic reliable. Redis helps solve the problem of when either the web application crashes or BJP itself crashes, the jobs stored in Redis will be persisted. In case of Redis crashing, we will also have the jobs persisted to disk thanks to the database snapshots taken by Redis.
+
+This solves the problem of the main application crashing or the BJP crashing. What should be done if the jobs themselves crash/raise an exception?
 
 #### Retrying Failed Jobs & Job Scheduler
+
+Jobs can fail for numerous reasons that may or may not be in the developer's control such as temporary network issues, timeouts, invalid job, etc. Regardless of the reason, the BJP needs to handle these errors. In case the error being raised while the job is being performed is a momentary error, then it might be a good idea to have a way to retry this job later in the future.
+
 ![Retry Failed Jobs Diagram](/images/job_retry.png)
 
-Jobs can also fail for numerous reasons that may or may not be in the developer's control such as temporary network issues, timeouts, invalid job, etc. Regardless of the reason, Workerholic will attempt to retry a job. The way we set up job retrying is to schedule it for some time in the future, effectively turning a failed job into a scheduled job. Here is a little bit of code to show you what that looks like:
+Workerholic will attempt to retry failed jobs. The way we set up job retrying is to schedule it for some time in the future, effectively turning a failed job into a scheduled job. Here is the implementation of Workerholic's `retry` functionality:
```

Comment:
"Workerholic will attempt to retry intermittently failing jobs. The way we set up job retrying is by scheduling the job to be performed at some time in the future, effectively turning a failed job into a scheduled job. Here is the implementation of Workerholic's `retry` functionality:"